nav2_waypoint_follower/src/waypoint_follower.cpp

```diff
@@ -304,4 +329,174 @@ WaypointFollower::goalResponseCallback(
+  {
+    RCLCPP_ERROR(get_logger(), "Send goal call failed");
+    return;
+  }
+  // Get the goal handle and save so that we can check
+  // on completion in the timer callback
+  waypoint_follower_goal_handle_ = future_goal_handle.get();
+  if (!waypoint_follower_goal_handle_) {
+    RCLCPP_ERROR(get_logger(), "Goal was rejected by server");
+    return;
+  }
+}
+
+std::vector<geometry_msgs::msg::PoseStamped>
+WaypointFollower::convertGPSWaypointstoPosesinMap(
+  const
+  std::vector<sensor_msgs::msg::NavSatFix> & gps_waypoints,
+  const rclcpp_lifecycle::LifecycleNode::SharedPtr & parent_node,
+  const rclcpp::Client<robot_localization::srv::FromLL>::SharedPtr & fromll_client)
+{
+  RCLCPP_INFO(parent_node->get_logger(), "Converting GPS waypoints to Map Frame..");
+  if (!parent_node) {
+    throw std::runtime_error{
+            "Failed to lock node while trying to convert GPS waypoints to Map frame"};
+  }
+  std::vector<geometry_msgs::msg::PoseStamped> poses_in_map_frame_vector;
+  int index_of_gps_waypoints = 0;
+  for (auto && curr_gps_waypoint : gps_waypoints) {
+    auto fromLLRequest = std::make_shared<robot_localization::srv::FromLL::Request>();
+    fromLLRequest->ll_point.latitude = curr_gps_waypoint.latitude;
+    fromLLRequest->ll_point.longitude = curr_gps_waypoint.longitude;
+    fromLLRequest->ll_point.altitude = curr_gps_waypoint.altitude;
+    if (!fromll_client->wait_for_service((std::chrono::seconds(10)))) {
+      RCLCPP_ERROR(
+        parent_node->get_logger(),
+        "fromLL service from robot_localization is not available"
+        "cannot convert GPS waypoints to Map frame poses,"
+        "Make sure you have run navsat_transform_node of robot_localization");
+      return std::vector<geometry_msgs::msg::PoseStamped>();
+    }
+    RCLCPP_INFO(
+      parent_node->get_logger(), "Processing for %i'th Waypoint..",
+      index_of_gps_waypoints);
+    auto inner_client_callback =
+      [&, parent_node](rclcpp::Client<robot_localization::srv::FromLL>::SharedFuture inner_future)
+      {
+        auto result = inner_future.get();
+      };
+    auto inner_future_result = fromll_client->async_send_request(
+      fromLLRequest,
+      inner_client_callback);
+    // this poses are assumed to be on global frame (map)
+    geometry_msgs::msg::PoseStamped curr_waypoint_in_map_frame;
+    curr_waypoint_in_map_frame.header.frame_id = "map";
+    curr_waypoint_in_map_frame.header.stamp = rclcpp::Clock().now();
+    curr_waypoint_in_map_frame.pose.position.x = inner_future_result.get()->map_point.x;
+    curr_waypoint_in_map_frame.pose.position.y = inner_future_result.get()->map_point.y;
+    curr_waypoint_in_map_frame.pose.position.z = inner_future_result.get()->map_point.z;
+
+    tf2::Quaternion quat_tf;
+    quat_tf.setRPY(0, 0, 0);
+    geometry_msgs::msg::Quaternion quat_msg;
+    tf2::convert(quat_msg, quat_tf);
+    curr_waypoint_in_map_frame.pose.orientation = quat_msg;
```

Comment:
I'm confused, why create a `quat_tf` and then convert it into a `quat_msg` to pass to the orientation without setting any orientation? You never used the results from the service to set the orientation. 

Also an aside, but actually `geometry_msgs::msg::Quaternion` defaults to a identity transform `(0 0 0 1)` now. No need for the rest of that setting to 0/0/0

Reply:
Actually, there is one thing we might clarify here. Since GPS data has no information of orientation, if we set orientation to zero at each waypoint the robot will behave quite not smart. IMO good option here will be set orientation to current robot orientation so recovery server wont intervene if robot is unable to reorient itself. We can use tf of map->base_link to get robot orientation but is there another simpler way ?

Reply:
I understand the issue you bring up (that we don't have orientation in GPS coords, so what orientation should we use?) but I don't understand the suggested solution. 

My suggestion would be to modify the message and have each GPS waypoint also define an orientation for the waypoint. Then that would trickle down to the WP follower to have an orientation to use. The question then is what frame is the orientation in (GPS?) and if GPS, how to convert it to map. This must be a common thing people do so I imagine there's conventions for it in GPS-land. 

I think that's the best solution, however for some users that don't fill in the waypoints completely, I agree having some reasonable default behavior if all the waypoint orientations are 0 is good, but not sure what the right option is. If you take the line segment created by wp `i` and wp `i+1`, find the map-frame orientation of the vector, and use that? That way on approach to the waypoint, the orientation will be the same as the approach vector.

Reply:
I have been thinking about this so it took some time for me to fully realize. To my understanding the IMU is represented in UMT coordinates.

Reply:
I think what you're saying is that the orientations can be transformed using the RL service, right? It should be pretty easy to test in gazebo if you set some angles you know what they should be and see if the robot ends up there in map frame.

Reply:
> using the RL service, right?

Yes but indirectly, because the broadcaster of `utm ->map` transform is RL so yeah basically we rely on RL here, that transform gives the goal's orientation in map frame, which robot understands correctly. I did a test in Gazebo, the approach seems to work. 
As an illutration; 
```
    wp0: [49.899954067243776, 8.899896379691963, 0.6342220147844688,1.57] 
    wp1: [49.899954067243776, 8.899896379691963, 0.6342220147844688, 3.14] 
``` 
Gazebo IMU report angle increases counterclockwise, the above waypoints positions are the same, robot reaches to first waypoint, heading 1.57 rad. left to origin of world, and then rotates 1.57 rad more , finally reaching correct orientation. This is was of course in fully perfect simulation environment, not sure how this will perform in real but we will see.

Reply:
That would be good for you to mention in the navigation.ros.org PR in migration guide / WP follower configuration page